core/src/main/scala/com/mesosphere/usi/core/matching/PortResource.scala

```diff
@@ -0,0 +1,223 @@
+    *   we hit the last offered port with wrap around and start offering the ports at the beginning
+    *   of the sequence up to (excluding) the port index we started at.
+    * * The next range is determined on demand. That's why an iterator is returned.
+    */
+  private def lazyRandomPortsFromRanges(offeredPortRanges: Seq[PortRange], rand: Random): Iterator[Int] = {
+    val numberOfOfferedPorts = offeredPortRanges.map(_.size).sum
+
+    if (numberOfOfferedPorts == 0) {
+      return Iterator.empty
+    }
+
+    def findStartPort(shuffled: IndexedSeq[PortRange], startPortIdx: Int): (Int, Int) = {
+      var startPortIdxOfCurrentRange = 0
+      val rangeIdx = shuffled.indexWhere {
+        case range: PortRange if startPortIdxOfCurrentRange + range.size > startPortIdx =>
+          true
+        case range: PortRange =>
+          startPortIdxOfCurrentRange += range.size
+          false
+      }
+
+      (rangeIdx, startPortIdx - startPortIdxOfCurrentRange)
+    }
+
+    val shuffled = rand.shuffle(offeredPortRanges).toIndexedSeq
+    val startPortIdx = rand.nextInt(numberOfOfferedPorts)
+    val (rangeIdx, portInRangeIdx) = findStartPort(shuffled, startPortIdx)
+    val startRangeOrig = shuffled(rangeIdx)
+
+    val startRange = startRangeOrig.drop(portInRangeIdx)
+
+    // These are created on demand if necessary
+    def afterStartRange: Iterator[Int] =
+      shuffled.slice(rangeIdx + 1, shuffled.length).iterator.flatMap(_.iterator)
+    def beforeStartRange: Iterator[Int] =
+      shuffled.slice(0, rangeIdx).iterator.flatMap(_.iterator)
+    def endRange: Iterator[Int] = startRangeOrig.take(portInRangeIdx)
+
+    startRange ++ afterStartRange ++ beforeStartRange ++ endRange
+  }
+}
+
+case class PortRange(minPort: Int, maxPort: Int) {
+  private[this] def range: Range.Inclusive = Range.inclusive(minPort, maxPort)
+  def size: Int = range.size
+
+  def iterator: Iterator[Int] = range.iterator.map(port => port)
+  def drop(n: Int): Iterator[Int] = range.drop(n).iterator.map(port => port)
+  def take(n: Int): Iterator[Int] = range.take(n).iterator.map(port => port)
+
+  /*
+   * Attention! range exports _two_ contains methods, a generic inefficient one and an efficient one
+   * that only gets used with Int (and not java.lang.Integer and similar)
+   */
+  def contains(port: Int): Boolean = range.contains(port)
+}
+
+sealed trait PortMatchResult
+case object PortNotAvailable extends PortMatchResult
+case class PortMatched(port: Int) extends PortMatchResult
+
+object PortResource {
+  val RandomPort: Int = 0
+
+  /**
+    * Return RangesResources covering all given ports with the given roles.
+    *
+    * Creates as few RangesResources as possible while
+    * preserving the order of the ports.
+    */
+  def createPortsResources(
+      originalResource: Protos.Resource,
+      requestedPorts: Seq[Int],
+      resourceType: ResourceType): Seq[Protos.Resource] = {
```

Comment:
Why do we need a `ResourceType` here if the method already says that it will create PortsResources?